code/game/machinery/turrets/turrets.dm

```diff
@@ -0,0 +1,225 @@
+	var/cell_type = /obj/item/cell/high // The type this turret will have on initialization.
+	var/can_charge_cell = TRUE // If FALSE, the battery has to be swapped out manually.
+	var/can_swap_cell = FALSE // If TRUE, the cell can be replaced with a screwdriver.
+	var/cell_looting_prob = 25 // Similar story as the gun looting var.
+
+	// Angles
+	// Remember that in BYOND, NORTH equals 0 absolute degrees, and not 90.
+	var/traverse = 180 // Determines how wide the turret can turn to shoot things, in degrees. The 'front' of the turret is determined by it's dir variable.
+	var/leftmost_traverse = null // How far left or right the turret can turn. Set automatically using the above variable and the inital dir value.
+	var/rightmost_traverse = null
+	var/current_bearing = 0 // Current absolute angle the turret has, used to calculate if it needs to turn to try to shoot the target.
+	var/target_bearing = 0 // The desired bearing. If the current bearing is too far from this, the turret will turn towards it until within tolerence.
+	var/bearing_tolerence = 5 // Degrees that the turret must be within to be able to shoot at the target.
+	var/turning_rate = 90 // Degrees per second.
+
+	// Detection.
+	var/datum/proximity_trigger/square/proximity = null
+	var/vision_range = 7	// How many tiles away the turret can see. Values higher than 7 will let the turret shoot offscreen, which might be unsporting.
+							// Higher values may also have a performance cost.
+	
+	// Logic
+	var/datum/state_machine/turret/state_machine = null
+	var/atom/target = null
+
+
+
+/obj/machinery/turret/Initialize()
+	state_machine = new(src)
+	proximity = new(src, 
+		/obj/machinery/turret/proc/on_proximity,
+		/obj/machinery/turret/proc/on_changed_turf_visibility,
+		vision_range,
+		PROXIMITY_EXCLUDE_HOLDER_TURF
+	)
+	proximity.register_turfs()
+	
+	internal_cell = new cell_type(src)
+	installed_gun = new gun_type(src)
+	target_bearing = dir2angle(dir)
+	set_bearing(target_bearing)
+	calculate_traverse()
+	update_icon()
+
+	START_PROCESSING(SSfastprocess, src) // TODO
+	return ..()
+
+/obj/machinery/turret/Destroy()
+	QDEL_NULL(state_machine)
+	QDEL_NULL(proximity)
+	if(istype(internal_cell))
+		QDEL_NULL(internal_cell)
+	if(istype(installed_gun))
+		QDEL_NULL(installed_gun)
+	STOP_PROCESSING(SSfastprocess, src) // TODO
+	return ..()
+
+/obj/machinery/turret/Process()
+	process_power()
+
+	state_machine.evaluate()
+	var/decl/state/turret/turret_state = state_machine.current_state
+	if(turret_state.type == /decl/state/turret/turning) // TODO remove after test.
+		process_turning()
+	else if(turret_state.type == /decl/state/turret/engaging)
+		process_shooting()
+
+/obj/machinery/turret/proc/process_power()
+	// If the turret is dead or turned off, don't draw any power.
+	if((stat & BROKEN) || !internal_cell)
+		update_use_power(POWER_USE_OFF)
+		// TODO Consider making turret stop processing later.
+		return
```

Comment:
Couple of `elses` here and below that are right after `returns` and aren't strictly needed.